Secondary indexing review (#1467)
redis/redis-doc: topics/indexes.md

# Secondary indexing with Redis

There are two general ways to create indexes on secondary attributes in Redis:

* Use native Redis data structures to index secondary attributes manually.
* Use the [RediSearch module](https://oss.redislabs.com/redisearch/) to create secondary indexes declaratively and query those indexes using a built-in query language.

> If you're interested in learning more about RediSearch, see the [official tutorial](https://github.com/RediSearch/redisearch-getting-started), the [RediSearch docs](https://oss.redislabs.com/redisearch/), or this [RediSearch video explainer](https://www.youtube.com/watch?v=B10nHEdW3NA).

Author: ```suggestion
To learn more about RediSearch, see the [official tutorial](https://github.com/RediSearch/redisearch-getting-started), the [RediSearch docs](https://oss.redislabs.com/redisearch/), or this [RediSearch video explainer](https://www.youtube.com/watch?v=B10nHEdW3NA).
```

If you want to build secondary indexes with native Redis data structures, then read on.

## Using Redis data structures to index data

It's fair to say that, natively, Redis only offers *primary key access*. However, since Redis is a data structures server, you can use the built-in data structures to create secondary indexes of different kinds, including composite (multi-column) indexes.

This document explains how it is possible to create indexes in Redis using the following data structures:

* Sorted sets to create secondary indexes by ID or other numerical fields.
* Sorted sets with lexicographical ranges for creating more advanced secondary indexes, composite indexes, and graph traversal indexes.
* Sets for creating random indexes.

> * Lists for creating simple iterable indexes and index on the  "last _N_ items*.

Author: ```suggestion
* Lists for creating simple iterable indexes and index on the  "last _N_ items".
```

    ZADD myindex 35 Jon
    ZADD myindex 67 Helen

To retrieve all persons with an age between 20 and 40, you can use the following command:

    ZRANGEBYSCORE myindex 20 40
    1) "Manuel"
    2) "Jon"

By using the **WITHSCORES** option of `ZRANGEBYSCORE`, you can also 
to obtain the scores associated with the returned elements.

You can also use the `ZCOUNT` command to get the number of elements
within a given range, without actually fetching the elements. 
This is especially useful because the operation is executed in logarithmic
time regardless of the size of the range.

Ranges can be inclusive or exclusive; please refer to the [`ZRANGEBYSCORE`
command documentation](/commands/zrangebyscore) for more information.

> **Note**: Using the `ZREVRANGEBYSCORE`, you can query a range in

Author: ```suggestion
Use `ZREVRANGEBYSCORE` to query ranges in reverse order.
```

> reverse order

Author: ```suggestion
```
